executor/partition_table.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// PartitionTableExecutor is a Executor for partitioned table.
+// it works by wrap the underlying TableReader/IndexReader/IndexLookUpReader.
+type PartitionTableExecutor struct {
+	baseExecutor
+
+	nextPartition
+	partitions []table.PhysicalTable
+	cursor     int
+	curr       Executor
+}
+
+type nextPartition interface {
+	nextPartition(context.Context, table.PhysicalTable) (Executor, error)
+}
+
+type nextPartitionForTableReader struct {
+	exec *TableReaderExecutor
+}
+
+func (n nextPartitionForTableReader) nextPartition(ctx context.Context, tbl table.PhysicalTable) (Executor, error) {
+	n.exec.table = tbl
+	return n.exec, nil
+}
+
+type nextPartitionForIndexLookUp struct {
+	exec *IndexLookUpExecutor
+}
+
+func (n nextPartitionForIndexLookUp) nextPartition(ctx context.Context, tbl table.PhysicalTable) (Executor, error) {
+	n.exec.table = tbl
+	return n.exec, nil
+}
+
+type nextPartitionForIndexReader struct {
+	exec *IndexReaderExecutor
+}
+
+func (n nextPartitionForIndexReader) nextPartition(ctx context.Context, tbl table.PhysicalTable) (Executor, error) {
+	exec := n.exec
+	exec.table = tbl
+	exec.physicalTableID = tbl.GetPhysicalID()
+	return exec, nil
+}
+
+func nextPartitionWithTrace(ctx context.Context, n nextPartition, tbl table.PhysicalTable) (Executor, error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan(fmt.Sprintf("nextPartition %d", tbl.GetPhysicalID()), opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		ctx = opentracing.ContextWithSpan(ctx, span1)
+	}
+	return n.nextPartition(ctx, tbl)
+}
+
+// Next implements the Executor interface.
+func (e *PartitionTableExecutor) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	var err error
+	for e.cursor < len(e.partitions) {
+		if e.curr == nil {
+			n := e.nextPartition
+			e.curr, err = nextPartitionWithTrace(ctx, n, e.partitions[e.cursor])
+			if err != nil {
+				return err
+			}
+			if err := e.curr.Open(ctx); err != nil {
+				return err
+			}
+		}
+
+		err = Next(ctx, e.curr, chk)
```

Comment:
maybe we need to take care `chk.IsFull()` question?

it seems current impl need read N partition for `limit 1`

Reply:
No, it's not read N partitions. You can verify that from the trace output:

```
mysql> set @try_new_partition_implementation = 1;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from p;
+------+------+
| id   | c    |
+------+------+
|    3 |    3 |
| NULL | NULL |
|    4 |    4 |
|    6 |    6 |
|    9 |    9 |
+------+------+
5 rows in set (0.00 sec)

mysql> trace format = 'row' select * from p where c > 0;
+------------------------------------------------+-----------------+------------+
| operation                                      | startTS         | duration   |
+------------------------------------------------+-----------------+------------+
| trace                                          | 20:46:22.883879 | 4.387835ms |
|   ├─session.Execute                            | 20:46:22.883892 | 984.726µs  |
|   │ ├─session.ParseSQL                         | 20:46:22.883908 | 52.458µs   |
|   │ └─session.ExecuteStmt                      | 20:46:22.884012 | 849.525µs  |
|   │   ├─executor.Compile                       | 20:46:22.884024 | 527.502µs  |
|   │   └─session.runStmt                        | 20:46:22.884599 | 209.83µs   |
|   ├─*executor.PartitionTableExecutor.Next      | 20:46:22.884951 | 1.092801ms |
|   │ ├─nextPartition 67                         | 20:46:22.884959 | 2.232µs    |
|   │ ├─TableReaderExecutor.Open                 | 20:46:22.884980 | 106.923µs  |
|   │ │ └─distsql.Select                         | 20:46:22.885018 | 51.34µs    |
|   │ │   └─regionReqauest.SendReqCtx            | 20:46:22.885221 | 603.301µs  |
|   │ └─*executor.TableReaderExecutor.Next       | 20:46:22.885126 | 880.063µs  |
|   ├─*executor.PartitionTableExecutor.Next      | 20:46:22.886069 | 971.84µs   |
|   │ ├─*executor.TableReaderExecutor.Next       | 20:46:22.886097 | 22.578µs   |
|   │ ├─nextPartition 68                         | 20:46:22.886164 | 2.476µs    |
|   │ ├─TableReaderExecutor.Open                 | 20:46:22.886183 | 119.504µs  |
|   │ │ └─distsql.Select                         | 20:46:22.886210 | 55.375µs   |
|   │ │   └─regionReqauest.SendReqCtx            | 20:46:22.886399 | 488.561µs  |
|   │ └─*executor.TableReaderExecutor.Next       | 20:46:22.886341 | 665.237µs  |
|   ├─*executor.PartitionTableExecutor.Next      | 20:46:22.887065 | 857.159µs  |
|   │ ├─*executor.TableReaderExecutor.Next       | 20:46:22.887069 | 16.232µs   |
|   │ ├─nextPartition 69                         | 20:46:22.887145 | 2.3µs      |
|   │ ├─TableReaderExecutor.Open                 | 20:46:22.887163 | 74.293µs   |
|   │ │ └─distsql.Select                         | 20:46:22.887189 | 30.45µs    |
|   │ │   └─regionReqauest.SendReqCtx            | 20:46:22.887307 | 451.362µs  |
|   │ └─*executor.TableReaderExecutor.Next       | 20:46:22.887264 | 623.905µs  |
|   └─*executor.PartitionTableExecutor.Next      | 20:46:22.887948 | 124.758µs  |
|     └─*executor.TableReaderExecutor.Next       | 20:46:22.887955 | 32.214µs   |
+------------------------------------------------+-----------------+------------+
28 rows in set (0.01 sec)

mysql> trace format = 'row' select * from p where c > 0 limit 1;
+---------------------------------------------------+-----------------+------------+
| operation                                         | startTS         | duration   |
+---------------------------------------------------+-----------------+------------+
| trace                                             | 20:46:17.283976 | 2.174428ms |
|   ├─session.Execute                               | 20:46:17.283985 | 983.808µs  |
|   │ ├─session.ParseSQL                            | 20:46:17.284000 | 54.235µs   |
|   │ └─session.ExecuteStmt                         | 20:46:17.284105 | 847.928µs  |
|   │   ├─executor.Compile                          | 20:46:17.284116 | 516.49µs   |
|   │   └─session.runStmt                           | 20:46:17.284668 | 228.442µs  |
|   ├─*executor.LimitExec.Next                      | 20:46:17.285002 | 980.05µs   |
|   │ └─*executor.PartitionTableExecutor.Next       | 20:46:17.285008 | 957.985µs  |
|   │   ├─nextPartition 67                          | 20:46:17.285014 | 2.305µs    |
|   │   ├─TableReaderExecutor.Open                  | 20:46:17.285033 | 112.509µs  |
|   │   │ └─distsql.Select                          | 20:46:17.285072 | 55.292µs   |
|   │   │   └─regionReqauest.SendReqCtx             | 20:46:17.285246 | 485µs      |
|   │   └─*executor.TableReaderExecutor.Next        | 20:46:17.285164 | 761.746µs  |
|   └─*executor.LimitExec.Next                      | 20:46:17.286006 | 2.443µs    |
+---------------------------------------------------+-----------------+------------+
14 rows in set (0.01 sec)
```

In the `limit 1` case, only the `nextPartition 67` is actually visited.